pkg/controller/constraintstatus/constraintstatus_controller.go

```diff
@@ -0,0 +1,248 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package constraintstatus
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/go-logr/logr"
+	opa "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/gatekeeper/apis/status/v1beta1"
+	"github.com/open-policy-agent/gatekeeper/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	"github.com/open-policy-agent/gatekeeper/pkg/watch"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	log = logf.Log.WithName("controller").WithValues(logging.Process, "constraint_status_controller")
+)
+
+type Adder struct {
+	Opa              *opa.Client
+	WatchManager     *watch.Manager
+	ControllerSwitch *watch.ControllerSwitch
+	Events           <-chan event.GenericEvent
+}
+
+func (a *Adder) InjectOpa(o *opa.Client) {
+	a.Opa = o
+}
+
+func (a *Adder) InjectWatchManager(w *watch.Manager) {
+}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {
+	a.ControllerSwitch = cs
+}
+
+// Add creates a new Constraint Status Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr, a.ControllerSwitch)
+	return add(mgr, r, a.Events)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(
+	mgr manager.Manager,
+	cs *watch.ControllerSwitch) reconcile.Reconciler {
+	return &ReconcileConstraintStatus{
+		// Separate reader and writer because manager's default client bypasses the cache for unstructured resources.
+		writer:       mgr.GetClient(),
+		statusClient: mgr.GetClient(),
+		reader:       mgr.GetCache(),
+
+		cs:     cs,
+		scheme: mgr.GetScheme(),
+		log:    log,
+	}
+}
+
+var _ handler.Mapper = &Mapper{}
+
+type Mapper struct {
+	packer util.EventPacker
+}
+
+func (m *Mapper) Map(obj handler.MapObject) []reconcile.Request {
```

Comment:
I'd love a comment that this maps from `ConstraintPodStatus` to their corresponding constraints.

Reply:
added